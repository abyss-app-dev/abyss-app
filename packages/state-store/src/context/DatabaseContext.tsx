import type { SQliteClient } from '@abyss/records';
import { createContext, type ReactNode, useContext } from 'react';

interface DatabaseContextType {
    database: SQliteClient | null;
}

const DatabaseContext = createContext<DatabaseContextType>({ database: null });

interface DatabaseProviderProps {
    database: SQliteClient;
    children: ReactNode;
}

export function DatabaseProvider({ database, children }: DatabaseProviderProps) {
    return <DatabaseContext.Provider value={{ database }}>{children}</DatabaseContext.Provider>;
}

export function useDatabase(): SQliteClient {
    const context = useContext(DatabaseContext);
    if (!context.database) {
        throw new Error('useDatabase must be used within a DatabaseProvider');
    }
    return context.database;
}
