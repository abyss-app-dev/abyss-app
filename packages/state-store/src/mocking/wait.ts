export function wait(ms = 50) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
