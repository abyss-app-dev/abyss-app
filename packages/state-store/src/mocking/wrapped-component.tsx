import type { StateRecorder } from './state-recorder';

export function TestRecorder<T>({ functionToTest, recorder }: { functionToTest: () => T; recorder: StateRecorder }) {
    recorder.record(functionToTest());
    return <>content</>;
}
