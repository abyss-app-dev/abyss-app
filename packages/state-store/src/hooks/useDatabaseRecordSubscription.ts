import type { SqliteTableRecordType, SqliteTables } from '@abyss/records';
import { useEffect } from 'react';
import { useDatabase } from '../context/DatabaseContext';
import { type Dependencies, useQuery } from './useQuery';

export type UseDatabaseRecordSubscription<T extends keyof SqliteTables> = ReturnType<typeof useDatabaseRecordSubscription<T>>;

export function useDatabaseRecordSubscription<T extends keyof SqliteTables>(
    table: T,
    recordId: string | undefined,
    dependencies: Dependencies = []
) {
    const database = useDatabase();

    // Query for our record itself
    const query = useQuery<SqliteTableRecordType[T] | null>(async () => {
        if (recordId) {
            return database.tables[table].get(recordId) as Promise<SqliteTableRecordType[T]>;
        }
        return null;
    }, [table, recordId, ...dependencies]);

    // Subscribe to our record itself and update our data on change
    useEffect(() => {
        if (!recordId) return;
        let unsubscribeCallback: () => void = () => {};
        database.tables[table]
            .subscribeRecord(recordId, data => query.setData(data as SqliteTableRecordType[T] | null))
            .then(unsubscribe => {
                unsubscribeCallback = unsubscribe;
            });
        return () => unsubscribeCallback();
    }, [database, table, recordId, query.setData, ...dependencies]);

    return query;
}
