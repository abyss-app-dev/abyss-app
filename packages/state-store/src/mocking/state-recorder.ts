import { render } from '@testing-library/react';
import type { ReactNode } from 'react';
import { expect } from 'vitest';
import { wait } from './wait';

export class StateRecorder {
    private states: unknown[] = [];

    record(state: unknown): void {
        this.states.push(state);
    }

    async delayedResolve(): Promise<unknown[]> {
        await wait();
        return this.states;
    }

    getStates(): unknown[] {
        return this.states;
    }

    getLatestState(): unknown {
        return this.states[this.states.length - 1];
    }

    clear(): void {
        this.states.length = 0;
    }

    async renderToSnapshot(component: ReactNode): Promise<void> {
        render(component);
        expect(await this.delayedResolve()).toMatchSnapshot();
    }
}
