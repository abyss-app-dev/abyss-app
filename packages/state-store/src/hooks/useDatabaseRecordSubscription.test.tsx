import { buildTestDB, type SQliteClient, SqliteTable } from '@abyss/records';
import { useEffect, useState } from 'react';
import { beforeEach, describe, test } from 'vitest';
import { DatabaseProvider } from '../context/DatabaseContext';
import { StateRecorder } from '../mocking/state-recorder';
import { TestRecorder } from '../mocking/wrapped-component';
import { useDatabaseRecordSubscription } from './useDatabaseRecordSubscription';

describe('useDatabaseRecordSubscription hook', () => {
    let db: SQliteClient;
    let recorder: StateRecorder;

    beforeEach(async () => {
        db = await buildTestDB();
        recorder = new StateRecorder();
    });

    test('subscribe to existing record with known ID', async () => {
        // Pre-create the record to get a known ID
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const subscription = useDatabaseRecordSubscription(SqliteTable.settings, defaultRecord.id);
                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscribe to non-existent record', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const subscription = useDatabaseRecordSubscription(SqliteTable.settings, 'settings::nonexistent');
                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('handle null record ID', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const subscription = useDatabaseRecordSubscription(SqliteTable.settings, undefined);
                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('record ID changes from null to valid', async () => {
        // Pre-create the record to get a known ID
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [recordId, setRecordId] = useState<string | undefined>(undefined);

                        useEffect(() => {
                            if (recordId === undefined) {
                                setTimeout(() => setRecordId(defaultRecord.id));
                            }
                        }, [recordId]);

                        const subscription = useDatabaseRecordSubscription(SqliteTable.settings, recordId);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('record ID changes from valid to null', async () => {
        // Pre-create the record to get a known ID
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [recordId, setRecordId] = useState<string | undefined>(defaultRecord.id);

                        useEffect(() => {
                            if (recordId === defaultRecord.id) {
                                setTimeout(() => setRecordId(undefined));
                            }
                        }, [recordId]);

                        const subscription = useDatabaseRecordSubscription(SqliteTable.settings, recordId);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription with dependencies that change', async () => {
        // Pre-create the record to get a known ID
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [dependency, setDependency] = useState(1);

                        useEffect(() => {
                            if (dependency === 1) {
                                setTimeout(() => setDependency(2));
                            }
                        }, [dependency]);

                        const subscription = useDatabaseRecordSubscription(SqliteTable.settings, defaultRecord.id, [dependency]);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('manually call refetch once', async () => {
        // Pre-create the record to get a known ID
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasRefetched, setHasRefetched] = useState(false);

                        const subscription = useDatabaseRecordSubscription(SqliteTable.settings, defaultRecord.id);

                        useEffect(() => {
                            if (subscription.data && !subscription.loading && !hasRefetched) {
                                setHasRefetched(true);
                                setTimeout(() => subscription.refetch());
                            }
                        }, [subscription.data, subscription.loading, subscription.refetch, hasRefetched]);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('manually call setData', async () => {
        // Pre-create the record to get a known ID
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasSetData, setHasSetData] = useState(false);

                        const subscription = useDatabaseRecordSubscription(SqliteTable.settings, defaultRecord.id);

                        useEffect(() => {
                            if (subscription.data && !subscription.loading && !hasSetData) {
                                setHasSetData(true);
                                setTimeout(() =>
                                    subscription.setData({
                                        id: defaultRecord.id,
                                        theme: 'manually-set-theme',
                                        lastPage: '/manual',
                                        createdAt: Date.now(),
                                        updatedAt: Date.now(),
                                    })
                                );
                            }
                        }, [subscription.data, subscription.loading, subscription.setData, hasSetData]);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });
});
