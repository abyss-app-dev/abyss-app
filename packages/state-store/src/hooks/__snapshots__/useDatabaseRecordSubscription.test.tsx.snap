// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useDatabaseRecordSubscription hook > handle null record ID 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecordSubscription hook > manually call refetch once 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092442,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092442,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092442,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092442,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092442,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092442,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092442,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092442,
    },
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092442,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092442,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecordSubscription hook > manually call setData 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092706,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092706,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092706,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092706,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092706,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092706,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092713,
      "id": "settings::default",
      "lastPage": "/manual",
      "theme": "manually-set-theme",
      "updatedAt": 1748759092713,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecordSubscription hook > record ID changes from null to valid 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759091650,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759091650,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759091650,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759091650,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecordSubscription hook > record ID changes from valid to null 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759091913,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759091913,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759091913,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759091913,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecordSubscription hook > subscribe to existing record with known ID 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759090860,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759090860,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759090860,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759090860,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecordSubscription hook > subscribe to non-existent record 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Record with id settings::nonexistent not found in table settings],
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecordSubscription hook > subscription with dependencies that change 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092180,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092180,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092180,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092180,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759092180,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759092180,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;
